doc/horizontal_autoscaling.md

```diff
@@ -0,0 +1,173 @@
+
+```yaml
+apiVersion: cloud.paddlepaddle.org/v1
+kind: TrainingJob
+metadata:
+  name: job-1
+spec:
+  trainer:
+    entrypoint: "python train.py"
+    workspace: "/home/job-1/"
+    min-replica: 3
+    max-replica: 6
+    resources:
+      limits:
+        alpha.kubernetes.io/nvidia-gpu: 1
+        cpu: "800m"
+        memory: "1Gi"
+      requests:
+        cpu: "500m"
+        memory: "600Mi"
+  pserver:
+    min-replica: 3
+    max-replica: 3
+    resources:
+      limits:
+        cpu: "800m"
+        memory: "1Gi"
+      requests:
+        cpu: "500m"
+        memory: "600Mi"
+```
+
+The training job controller will create and continuously scale the
+number of trainers and the number of pservers between the
+corresponding `min-replica` and `max-replica`.
+
+The training job custom resource can be created with: `kubectl create
+-f training_job.yaml`.
+
+The custom resource will only be saved on Kuberentes, we will need a
+custom controller that operate on it.
+
+### Training Job Controller
+
+The training job controller run as a Pod. It has the global view of
+the computation resources. It watches the training job resources and
+schedules and scales the training jobs using the Kuberenetes API.
+
+The pseudo controller declaration (`training_job_controller.yaml`) is
+as follows:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: training-job-controller
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        name: training-job-controller
+    spec:
+      containers:
+      - name: training-job-controller
+        image: paddlepaddle/training-job-controller
+```
+
+The training job controller can be started by the cluster
+administrator with command: `kubectl create -f
+training_job_controller.yaml`
+
+## Implementation
+
+### Training Job Resource
+
+The training job resource is a custom resource, there are two ways of
+implementing custom resources:
+
+- [Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/),
+  since Kuberentes v1.7.
+- [Third Party Resource (TPR)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/),
+  since Kubernetes v1.2, fully deprecated in v1.8, will be removed in
+  v1.9.
+
+We will be using CRD unless our cluster can not be updated to
+Kubernetes v1.7, since TPR will be removed in v1.9.
+
+### Training Job Controller
+
+Currently, we will run a single training job instance and assume that
+there is no training job controller running concurrently (the
+assumption could be false when split-brain happens). In the future, we
+will run multiple instances and use leader election to choose a
+leader.
+
+The pseudo-logic is as follows:
+
+```go
+registerThirdPartyResource()
+for {
+  quota := getTotalComputationResourceQuota()
+  current := getCurrentJobStates()
+  desired := getDesiredJobStates()
+  dynamicSchedule(quota, current, desired)
```

Comment:
If there are many of tasks and the quotas can't meet all of them, who's need will be met first?

Reply:
Great question! Added "Scaling Algorithm" section, please take a look.